File: core-api-backend/api/routes/websocket.ts
import { Router } from 'express';
import { Server as HttpServer } from 'http';
import { WebSocket, WebSocketServer } from 'ws';

const router = Router();

// Track active WebSocket connections
const clients = new Set<WebSocket>();

// Track the ping interval
let pingInterval: NodeJS.Timeout | null = null;

// Initialize WebSocket server
export function initWebSocket(server: HttpServer) {
  const wss = new WebSocketServer({ server });

  wss.on('connection', (ws: WebSocket) => {
    // Add client to active connections
    clients.add(ws);
    console.log('WebSocket client connected - total clients:', clients.size);
    
    // Send initial connection message
    ws.send(JSON.stringify({ 
      type: 'connection', 
      message: 'Connected to WebSocket server' 
    }));
    
    // Start the ping interval if needed
    startPingInterval();

    // Handle incoming messages from clients
    ws.on('message', (message: string) => {
      try {
        const data = JSON.parse(message.toString());
        console.log('Received message:', data);
        
        // You can handle specific message types here
        // For example, echo back the message
        ws.send(JSON.stringify({ 
          type: 'echo', 
          message: data,
          timestamp: new Date().toISOString()
        }));
      } catch (error) {
        console.error('Error parsing message:', error);
      }
    });

    // Handle connection closing
    ws.on('close', () => {
      clients.delete(ws);
      console.log('WebSocket client disconnected - remaining clients:', clients.size);
      
      // Stop the ping interval if no clients are connected
      stopPingInterval();
    });

    // Handle errors
    ws.on('error', (error) => {
      console.error('WebSocket error:', error);
      clients.delete(ws);
    });
  });

  return wss;
}

// Start the ping interval if it's not already running
function startPingInterval() {
  if (pingInterval === null && clients.size > 0) {
    console.log('Starting ping interval - clients connected:', clients.size);
    pingInterval = setInterval(() => {
      sendToAll({ 
        type: 'ping', 
        message: 'Server heartbeat', 
        timestamp: new Date().toISOString() 
      });
    }, 1000);
  }
}

// Stop the ping interval if no clients are connected
function stopPingInterval() {
  if (pingInterval !== null && clients.size === 0) {
    console.log('Stopping ping interval - no clients connected');
    clearInterval(pingInterval);
    pingInterval = null;
  }
}

// Helper function to send message to all connected clients
export const sendToAll = (data: any) => {
  const message = JSON.stringify(data);
  clients.forEach(client => {
    if (client.readyState === WebSocket.OPEN) {
      client.send(message);
    }
  });
};

// This endpoint can be used to check WebSocket status
router.get('/status', (req, res) => {
  res.json({
    active: clients.size,
    status: 'running'
  });
});

export default router;

File: core-api-backend-py/app/api/routes/websocket.py
import asyncio
import json
from datetime import datetime
from typing import List, Dict, Any
import logging

# Configure basic logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)

from fastapi import APIRouter, WebSocket, WebSocketDisconnect

router = APIRouter()

# Track active WebSocket connections
active_connections: List[WebSocket] = []

# Track the ping task
ping_task = None

async def send_to_all(data: Dict[str, Any]):
    """Send a message to all connected clients"""
    message = json.dumps(data)
    for connection in active_connections:
        try:
            await connection.send_text(message)
        except Exception as e:
            logging.error(f"Error sending message: {e}")

async def ping_clients():
    """Send a ping to all clients every second"""
    while True:
        if active_connections:
            data = {
                "type": "ping",
                "message": "Server heartbeat",
                "timestamp": datetime.utcnow().isoformat()
            }
            await send_to_all(data)
            logging.info(f"Sent ping to {len(active_connections)} websocket clients")
        await asyncio.sleep(1)

# Define the WebSocket endpoint that will be used in main.py
async def websocket_endpoint(websocket: WebSocket):
    """Handle WebSocket connections"""
    await websocket.accept()
    active_connections.append(websocket)
    
    # Start ping task if not already running
    global ping_task
    if ping_task is None or ping_task.done():
        ping_task = asyncio.create_task(ping_clients())
    
    logging.info(f"WebSocket client connected - total clients: {len(active_connections)}")
    
    # Send initial connection message
    await websocket.send_json({
        "type": "connection",
        "message": "Connected to WebSocket server"
    })
    
    try:
        while True:
            # Receive message from the client
            data = await websocket.receive_text()
            
            try:
                # Parse the message
                parsed_data = json.loads(data)
                logging.info(f"Received message: {parsed_data}")
                
                # Echo back the message
                await websocket.send_json({
                    "type": "echo",
                    "message": parsed_data,
                    "timestamp": datetime.utcnow().isoformat()
                })
            except json.JSONDecodeError as e:
                logging.error(f"Error parsing message: {e}")
                await websocket.send_json({
                    "type": "error",
                    "message": f"Invalid JSON: {str(e)}"
                })
    except WebSocketDisconnect:
        # Remove connection when client disconnects
        if websocket in active_connections:
            active_connections.remove(websocket)
        logging.info(f"WebSocket client disconnected - remaining clients: {len(active_connections)}")
    except Exception as e:
        logging.error(f"WebSocket error: {e}")
        if websocket in active_connections:
            active_connections.remove(websocket)

# Add a route for testing WebSocket functionality via HTTP
@router.get("/status")
async def websocket_status():
    """Get WebSocket connection status"""
    return {
        "active": len(active_connections),
        "status": "running"
    }